src/process/init.rs

```diff
@@ -249,18 +260,34 @@ pub fn container_init(args: ContainerInitArgs) -> Result<()> {
     };
 
     // clean up and handle perserved fds.
-    cleanup_file_descriptors(preserve_fds).with_context(|| "Failed to clean up extra fds")?;
+    if args.init {
+        cleanup_file_descriptors(preserve_fds).with_context(|| "Failed to clean up extra fds")?;
+    }
+
+    // Reset the process env based on oci spec.
+    env::vars().for_each(|(key, _value)| std::env::remove_var(key));
+    utils::parse_env(&envs)
+        .iter()
+        .for_each(|(key, value)| env::set_var(key, value));
 
     // notify parents that the init process is ready to execute the payload.
     child.notify_parent()?;
 
     // listing on the notify socket for container start command
     notify_socket.wait_for_container_start()?;
 
+    // create_container hook needs to be called after the namespace setup, but
+    // before pivot_root is called. This runs in the container namespaces.
+    if args.init {
+        if let Some(hooks) = hooks {
+            hooks::run_hooks(hooks.start_container.as_ref(), container)?
+        }
+    }
+
     if let Some(args) = proc.args.as_ref() {
-        utils::do_exec(&args[0], args, &envs)?;
+        utils::do_exec(&args[0], args)?;
     } else {
-        log::warn!("The command to be executed isn't set")
+        bail!("The command to be executed isn't set")
```

Comment:
Can I have the error message include the fact that args are required for non-Windows?
https://github.com/opencontainers/runtime-spec/blob/master/config.md#process